src/rvsdg/egglog_optimizer/rvsdg-logic.egg

```diff
@@ -0,0 +1,128 @@
+(datatype Situation
+    (always)
+    ; For last-iter, the Args of the predicate are bound in the Theta's context!
+    (last-iter Operand VecOperand VecOperand)
+    
+    ; Some more situations we may want in the future:
+    ; (in-context Context)
+    ; ; For kth-iter, the Args of the predicate are bound in the surrounding
+    ; ; context!
+    ; ;         kth iter of this Theta  Theta's *outer* context    
+    ; (kth-iter Body                    Context)
+
+    ; Note that all situations define the most recent binding body. We may want
+    ; situation to actually be a stack! E.g., we may want to reason about
+    ; an arg's value the kth iteration of a Theta, within a particular Gamma
+    ; branch 
+    )
+
+(datatype Predicate
+    ; [[o1]]_{s1} <= [[o2]]_{s2}
+    ;    s1        o1      s2        o2
+    (leq Situation Operand Situation Operand)
+    ; We prefer to write eq with two leq's because it makes rules simpler
+    ; Some other predicates we may want:
+    ; (eq-mod-k i64 Situation Operand Situation Operand)
+    )
+
+
+(relation might-create-situation (Body Situation))
+(rule ((= body (Gamma pred inputs outputs)))
+      ((might-create-situation body (always)))
+      :ruleset fast-analyses)
+(rule ((= body (Theta pred inputs outputs)))
+      ((might-create-situation body (always)))
+      :ruleset fast-analyses)
+(rule ((= situation (last-iter pred inputs outputs))
+       (= body (Theta pred inputs outputs)))
+      ((might-create-situation body situation))
+      :ruleset fast-analyses)
+
+(relation holds (Predicate))
+
+; leq situation weakening. There are more general ways to write this rule, but
+; we probably mainly want the situations to match
+(rule ((holds (leq s o1 (always) o2)))
+      ((holds (leq s o1 s o2)))
+      :ruleset fast-analyses)
+(rule ((holds (leq (always) o1 s o2)))
+      ((holds (leq s o1 s o2 )))
+      :ruleset fast-analyses)
+
+(relation demand-weakening (Predicate Situation))
+; TODO: see if a predicate can match a particular situation 
+
+; a constant has the same value in any situation
+(rule ((= c (Node (PureOp (Const typ cops lit))))
+       (holds (leq s1 o1 s2       c)))
+      ((holds (leq s1 o1 (always) c)))
+      :ruleset fast-analyses)
+(rule ((= c (Node (PureOp (Const typ cops lit))))
+       (holds (leq s1       c s2 o2)))
+      ((holds (leq (always) c s2 o2)))
+      :ruleset fast-analyses)
+
+; leq transitivity
+(rule ((holds (leq sa a sb b))
```

Comment:
Do you actually need this rule for this PR? Doesn't transitivity blow up?

Reply:
It's not needed. It doesn't blow up in this case since there are only small set of situations that can interrelate.

I commented it out regardless, but see my broader comment on blowup on your main review.